components/tidb_query_datatype/src/codec/collation/collator/gbk_bin.rs

```diff
@@ -0,0 +1,327 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use super::*;
+
+/// Collator for `gbk_bin` collation with padding behavior (trims right spaces).
+#[derive(Debug)]
+pub struct CollatorGBKBin;
+
+impl Collator for CollatorGBKBin {
+    // the charset of gbk_bin collation should be gbk, but actually we store the data with utf8mb4 character encoding, so mark its character to be utf8mb4.
+    type Charset = CharsetUtf8mb4;
+    type Weight = u16;
+
+    const IS_CASE_INSENSITIVE: bool = false;
+
+    #[inline]
+    fn char_weight(ch: char) -> Self::Weight {
+        // All GBK code point are in BMP, if the incoming character is not, convert it to '?'.
+        // This should not happened.
+        let r = ch as usize;
+        if r > 0xFFFF {
+            return '?' as u16;
+        }
+
+        GBK_BIN_TABLE[r]
+    }
+
+    #[inline]
+    fn write_sort_key<W: BufferWriter>(writer: &mut W, bstr: &[u8]) -> Result<usize> {
+        let s = str::from_utf8(bstr)?.trim_end_matches(PADDING_SPACE);
+        let mut n = 0;
+        for ch in s.chars() {
+            let weight = Self::char_weight(ch);
+            if weight > 0xFF {
+                writer.write_u16_be(weight)?;
+                n += 2;
+            } else {
+                writer.write_u8(weight as u8)?;
+                n += 1;
+            }
+        }
+        Ok(n * std::mem::size_of::<u8>())
+    }
+
+    #[inline]
+    fn sort_compare(a: &[u8], b: &[u8]) -> Result<Ordering> {
+        let sa = str::from_utf8(a)?.trim_end_matches(PADDING_SPACE);
+        let sb = str::from_utf8(b)?.trim_end_matches(PADDING_SPACE);
+        Ok(sa
+            .chars()
+            .map(Self::char_weight)
+            .cmp(sb.chars().map(Self::char_weight)))
+    }
+
+    #[inline]
+    fn sort_hash<H: Hasher>(state: &mut H, bstr: &[u8]) -> Result<()> {
+        let s = str::from_utf8(bstr)?.trim_end_matches(PADDING_SPACE);
+        for ch in s.chars().map(Self::char_weight) {
+            ch.hash(state);
+        }
+        Ok(())
+    }
+}
+
+// GBK_BIN_TABLE are the encoding tables from Unicode to GBK code, it is totally the same with golang's GBK encoding.
+// If there is no mapping code in GBK, use 0x3F(?) instead. It should not happened.
+#[rustfmt::skip]
+static GBK_BIN_TABLE: [u16; 0xffff + 1] = [
```

Comment:
Consider using `include_bytes!` macro, or using an existing library like https://docs.rs/encoding_rs/0.8.28/encoding_rs/static.GBK.html

I don't quite like embedding a such a large table in Rust code.

Reply:
- `include_bytes!` macro return `[u8; N]`, the table we need is `[u16; N]`, we need a special process to use it.
- we need to make sure the result is exactly the same as tidb which uses the golang GBK library, and I test the rust GBK library there is some difference that makes us cannot use it.

Reply:
... and writing such big array in Rust this will slower compile speed https://github.com/rust-lang/rust/issues/39352

Reply:
When I paste the full table in my local playground, the first thing to happen is that, my VSCode is stuck. No-one else will be willing to modify this part if it slows down the editor. So please at least create a separate file for that.

Regarding the compile speed,

```bash
cargo build
   Compiling rust-playground v0.1.0 (/home/skyzh/Work/rust-playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.84s
```

it seems not to be a big overhead.

Reply:
yes, large table makes editor stocks. I decided to use `include_bytes!`, PTAL